src/import/kv_importer.rs

```diff
@@ -0,0 +1,337 @@
+
+use std::fs;
+use std::fmt;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use uuid::Uuid;
+use kvproto::importpb::*;
+
+use config::DbConfig;
+use util::collections::HashMap;
+
+use super::engine::*;
+use super::{Config, Error, Result};
+
+pub type Token = usize;
+
+struct Inner {
+    engines: HashMap<Uuid, Arc<EngineFile>>,
+    clients: HashMap<Token, Arc<EngineFile>>,
+}
+
+pub struct KVImporter {
+    dir: EngineDir,
+    token: AtomicUsize,
+    inner: Mutex<Inner>,
+}
+
+impl KVImporter {
+    pub fn new(cfg: Config, opts: DbConfig) -> Result<KVImporter> {
+        let dir = EngineDir::new(&cfg.import_dir, opts)?;
+        Ok(KVImporter {
+            dir: dir,
+            token: AtomicUsize::new(1),
+            inner: Mutex::new(Inner {
+                engines: HashMap::default(),
+                clients: HashMap::default(),
+            }),
+        })
+    }
+
+    pub fn token(&self) -> Token {
+        self.token.fetch_add(1, Ordering::SeqCst)
+    }
+
+    fn insert(&self, token: Token, engine: Arc<EngineFile>) {
+        let mut inner = self.inner.lock().unwrap();
+        assert!(inner.clients.insert(token, engine).is_none());
+    }
+
+    pub fn remove(&self, token: Token) -> Option<Arc<EngineFile>> {
+        let mut inner = self.inner.lock().unwrap();
+        inner.clients.remove(&token)
+    }
+
+    pub fn open(&self, uuid: Uuid) -> Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        if inner.engines.contains_key(&uuid) {
+            return Ok(());
+        }
+
+        match self.dir.open(uuid) {
+            Ok(engine) => {
+                info!("open {:?}", engine);
+                inner.engines.insert(uuid, Arc::new(engine));
+                Ok(())
+            }
+            Err(e) => {
+                error!("open {}: {:?}", uuid, e);
+                Err(e)
+            }
+        }
+    }
+
+    /// Attach to the engine.
+    /// Clients with the same uuid will share the same engine.
+    pub fn attach(&self, token: Token, uuid: Uuid) -> Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        if inner.clients.contains_key(&token) {
+            return Err(Error::TokenExists(token));
+        }
+
+        let engine = match inner.engines.get(&uuid) {
+            Some(engine) => Arc::clone(engine),
+            None => return Err(Error::EngineNotFound(uuid)),
+        };
+
+        info!("attach {:?}", engine);
+        inner.clients.insert(token, engine);
+        Ok(())
+    }
+
+    pub fn detach(&self, token: Token) -> Result<()> {
+        match self.remove(token) {
+            Some(engine) => {
+                info!("detach {:?}", engine);
+                Ok(())
+            }
+            None => Err(Error::TokenNotFound(token)),
+        }
+    }
+
+    pub fn write(&self, token: Token, batch: WriteBatch) -> Result<usize> {
+        match self.remove(token) {
```

Comment:
It's wired that write contains a detach operation.

Reply:
We remove it here so we don't need to lock when writing to the engine.

Reply:
Is it an implication that a token can be only used to write once?

Reply:
No, a token can be used to write multiple times, that's why we insert it back after a successful write.